src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private int size;
+    private T[] array;
+
+    public ArrayList(int capacity) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("Array's capacity cannot be <= 0");
+        }
+        array = (T[]) new Object[capacity];
+    }
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index cannot be < 0 or > size");
+        }
+        checkCapacity();
+        if (size - index >= 0) {
+            System.arraycopy(array, index, array, index + 1, size - index);
+        }
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedElement = (T) array[index];
+        if (index < size - 1) {
+            System.arraycopy(array, index + 1, array, index, size - index - 1);
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((array[i] == null && element == null)
+                    || (array[i] != null && array[i].equals(element))) {
+                return remove(i);
```

Comment:
redundant parentheses, although you can keep them for the sake of better readability
condition can be simplified
```suggestion
            if (array[i] == element
                    || array[i] != null && array[i].equals(element)) {
                return remove(i);
```